### ISSUE TITLE
Consider using .Equals() to compare items in IReferenceOrISignature

### ISSUE THREAD

**Issue by cston** [dotnet/roslyn#48929]
In some scenarios, the dictionary `TokenMap._itemIdentityToToken` may contain a large number of entries that are essentially duplicate instances of the same C# symbol. For instance, in the test `OverloadResolutionPerfTests.ArrayInitializationAnonymousTypes`, each anonymous type instance has a unique `SubstitutedMethodSymbol` in `_itemIdentityToToken`, even though the method symbols are instances of the same anonymous type constructor.

Should `IReferenceOrISignature.Equals()` be changed to compare the underlying `_item` with `Equals()` rather than `ReferenceEquals()`? (It looks like that change does not break any existing tests.)

**Comment by jaredpar** [comment]
@AlekseyTs @cston 

From the referenced PR:

> In the bug repro, many symbols are added to the map for the same method (a SubstitutedMethodSymbol for an anonymous type constructor). ... There is a separate question of whether TokenMap should contain separate entries for multiple instances of the same symbol. I will confirm.

I think that is the central question here: is the expectation that the same symbol should have different entries in the token map here? If so then the current implementation seems correct. If not though then I think we should persue this route of doing proper equality checks.

**Comment by jaredpar** [comment]
Do we have a write up somewhere or comment that explains what "Fake" tokens are? It seems like we use them in a very real way in certain cases. Example [here](http://sourceroslyn.io/#Microsoft.CodeAnalysis/CodeGen/ILBuilderEmit.cs,46)

```cs
uint token = module?.GetFakeStringTokenForIL(value) ?? 0xFFFF;
this.GetCurrentWriter().WriteUInt32(token);
```

**Comment by AlekseyTs** [comment]
I do not believe there is a need for TokeMap to use any special equality other than reference equality. There is no requirement like that for correctness. Any attempt to try to infer tokens from symbols is error prone and I removed it in https://github.com/dotnet/roslyn/pull/48838.

**Comment by AlekseyTs** [comment]
Some symbols are going to be unified in the translation layer, the change is in the same PR

**Comment by AlekseyTs** [comment]
> A straight line reading of that code seems wrong. We're getting a fake token then writing it to IL. Feel like I'm missing a connection here.

Real tokens are assigned when we are emitting metadata. At that time we visit IL and replace "fake" numbers with real ones. Those are assigned based on the underlying Cci entities and the metadata generation layer knows when it can reuse real tokens and when it cannot. Alternatively, we will have to reimplement all that logic in the TokenMap comparer, I desided to not go that route until we have a real reason.

**Comment by AlekseyTs** [comment]
> Should IReferenceOrISignature.Equals() be changed to compare the underlying _item with Equals() rather than ReferenceEquals()? (It looks like that change does not break any existing tests.)

This suggestion goes against the change made in https://github.com/dotnet/roslyn/pull/48838

**Comment by jaredpar** [comment]
> Real tokens are assigned when we are emitting metadata. At that time we visit IL and replace "fake" numbers with real ones.

Where does this happen? Think seeing this code will help me understand the understanding gap I have here.

**Comment by AlekseyTs** [comment]
>> Real tokens are assigned when we are emitting metadata. At that time we visit IL and replace "fake" numbers with real ones.

> Where does this happen?

Microsoft.Cci.MetadataWriter.WriteInstructions
